server/src/main/java/org/elasticsearch/cluster/coordination/NodeJoinExecutor.java

```diff
@@ -42,10 +43,16 @@ public class NodeJoinExecutor implements ClusterStateTaskExecutor<JoinTask> {
 
     private final AllocationService allocationService;
     private final RerouteService rerouteService;
+    private final Function<ClusterState, ClusterState> maybeReconfigureAfterMasterElection;
 
-    public NodeJoinExecutor(AllocationService allocationService, RerouteService rerouteService) {
+    public NodeJoinExecutor(
+        AllocationService allocationService,
+        RerouteService rerouteService,
+        Function<ClusterState, ClusterState> maybeReconfigureAfterMasterElection
```

Comment:
```suggestion
        Function<ClusterState, ClusterState> maybeReconfigureAfterMasterElection
```

nit: I do like the explicitness of requiring this constructor parameter, but also those `Function.identity()` args add a lot of noise here. Maybe let's keep the 2-arg constructor here too and we can remove it in a follow-up.